File: python/elite_dangerous_sdk/query.py
"""Journal Event Query Engine - LINQ-style queries over journal events."""
from typing import Any, Callable, Optional


class JournalQuery:
    def __init__(self, events: list[dict]):
        self._results = list(events)

    def where(self, field_or_pred, value: Any = None, operator: str = "="):
        if callable(field_or_pred):
            self._results = [e for e in self._results if field_or_pred(e)]
            return self
        field = field_or_pred

        def _match(e: dict) -> bool:
            actual = e.get(field)
            if operator == "=":
                return actual == value
            elif operator == "!=":
                return actual != value
            elif operator == ">":
                return isinstance(actual, (int, float)) and isinstance(value, (int, float)) and actual > value
            elif operator == ">=":
                return isinstance(actual, (int, float)) and isinstance(value, (int, float)) and actual >= value
            elif operator == "<":
                return isinstance(actual, (int, float)) and isinstance(value, (int, float)) and actual < value
            elif operator == "<=":
                return isinstance(actual, (int, float)) and isinstance(value, (int, float)) and actual <= value
            return True

        self._results = [e for e in self._results if _match(e)]
        return self

    def between(self, start: str, end: str):
        self._results = [e for e in self._results if start <= e.get("timestamp", "") <= end]
        return self

    def select(self, field: str) -> list:
        return [e.get(field) for e in self._results]

    def count(self) -> int:
        return len(self._results)

    def first(self) -> Optional[dict]:
        return self._results[0] if self._results else None

    def last(self) -> Optional[dict]:
        return self._results[-1] if self._results else None

    def for_each(self, fn: Callable[[dict], Any]):
        for e in self._results:
            fn(e)

    def to_list(self) -> list[dict]:
        return list(self._results)


def query(events: list[dict]) -> JournalQuery:
    return JournalQuery(events)


def count_where(events: list[dict], event_type: str) -> int:
    return sum(1 for e in events if e.get("event") == event_type)


def filter_where(events: list[dict], event_type: str) -> list[dict]:
    return [e for e in events if e.get("event") == event_type]


def count_by_type(events: list[dict]) -> dict[str, int]:
    counts: dict[str, int] = {}
    for e in events:
        et = e.get("event", "")
        counts[et] = counts.get(et, 0) + 1
    return counts

File: core/packages/journal/src/query.ts
import type { JournalEvent } from "./types.js";

export class JournalQuery {
  private results: JournalEvent[];

  constructor(events: JournalEvent[]) {
    this.results = [...events];
  }

  where(field: string, value: unknown): JournalQuery;
  where(
    field: string,
    value: unknown,
    operator: "=" | "!=" | ">" | ">=" | "<" | "<=",
  ): JournalQuery;
  where(predicate: (event: JournalEvent) => boolean): JournalQuery;
  where(
    fieldOrPred: string | ((event: JournalEvent) => boolean),
    value?: unknown,
    operator: "=" | "!=" | ">" | ">=" | "<" | "<=" = "=",
  ): JournalQuery {
    if (typeof fieldOrPred === "function") {
      this.results = this.results.filter(fieldOrPred);
      return this;
    }
    this.results = this.results.filter((e) => {
      const actual = (e as Record<string, unknown>)[fieldOrPred];
      switch (operator) {
        case "=":
          return actual === value;
        case "!=":
          return actual !== value;
        case ">":
          return (
            typeof actual === "number" &&
            typeof value === "number" &&
            actual > value
          );
        case ">=":
          return (
            typeof actual === "number" &&
            typeof value === "number" &&
            actual >= value
          );
        case "<":
          return (
            typeof actual === "number" &&
            typeof value === "number" &&
            actual < value
          );
        case "<=":
          return (
            typeof actual === "number" &&
            typeof value === "number" &&
            actual <= value
          );
        default:
          return true;
      }
    });
    return this;
  }

  between(start: string, end: string): JournalQuery {
    this.results = this.results.filter(
      (e) => e.timestamp >= start && e.timestamp <= end,
    );
    return this;
  }

  select<T = unknown>(field: string): T[] {
    return this.results.map((e) => (e as Record<string, unknown>)[field] as T);
  }

  count(): number {
    return this.results.length;
  }

  first(): JournalEvent | undefined {
    return this.results[0];
  }

  last(): JournalEvent | undefined {
    return this.results[this.results.length - 1];
  }

  forEach(fn: (event: JournalEvent) => void): void {
    this.results.forEach(fn);
  }

  toArray(): JournalEvent[] {
    return [...this.results];
  }
}

export function query(events: JournalEvent[]): JournalQuery {
  return new JournalQuery(events);
}

export function countWhere(events: JournalEvent[], eventType: string): number {
  return events.filter((e) => e.event === eventType).length;
}

export function filterWhere<T extends JournalEvent = JournalEvent>(
  events: JournalEvent[],
  eventType: string,
): T[] {
  return events.filter((e) => e.event === eventType) as T[];
}

export function countByType(events: JournalEvent[]): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const e of events) {
    counts[e.event] = (counts[e.event] || 0) + 1;
  }
  return counts;
}
